File: imdb_search.py
import sys
import re
import json
from urllib.parse import quote_plus

try:
    import requests
    from bs4 import BeautifulSoup
    from PyQt6.QtWidgets import QApplication, QMainWindow, QVBoxLayout, QHBoxLayout, QWidget, QLineEdit, QPushButton, QLabel, QTextEdit
    from PyQt6.QtCore import Qt
    from PyQt6.QtGui import QClipboard
except ImportError:
    print("Error: Required libraries are not installed.")
    print("Please install the required libraries using the following command:")
    print("pip install requests beautifulsoup4 PyQt6")
    sys.exit(1)

def search_imdb_api(query, year=None):
    base_url = "https://v3.sg.media-imdb.com/suggestion/x/"
    search_url = f"{base_url}{quote_plus(query)}.json"
    
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
    }
    
    try:
        response = requests.get(search_url, headers=headers)
        response.raise_for_status()
        data = response.json()
        
        if 'd' in data:
            for item in data['d']:
                if item.get('q') in ['TV series', 'feature', 'TV mini-series']:
                    title = item.get('l', '').lower()
                    item_year = item.get('y')
                    
                    if title == query.lower() and (not year or str(item_year) == year):
                        return item.get('id')
        
        return None
    except requests.RequestException as e:
        print(f"API Error: Unable to connect to IMDB API. {str(e)}")
        return None

def search_imdb_web(query, year=None):
    encoded_query = quote_plus(query)
    search_url = f"https://www.imdb.com/search/title/?title={encoded_query}"
    if year:
        search_url += f"&release_date={year},{year}"

    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
    }
    
    try:
        response = requests.get(search_url, headers=headers)
        response.raise_for_status()
        soup = BeautifulSoup(response.content, 'html.parser')
        
        results = soup.find_all('div', class_='lister-item mode-advanced')
        
        for result in results:
            title_element = result.find('h3', class_='lister-item-header').find('a')
            title = title_element.text.strip()
            href = title_element['href']
            
            year_element = result.find('span', class_='lister-item-year')
            result_year = re.search(r'\d{4}', year_element.text).group() if year_element else None
            
            if title.lower() == query.lower() and (not year or year == result_year):
                match = re.search(r'/title/(tt\d+)/', href)
                if match:
                    return match.group(1)
        
        return None
    except requests.RequestException as e:
        print(f"Web Error: Unable to connect to IMDB. {str(e)}")
        return None

def search_imdb(query):
    print(f"Searching for: {query}")
    year_match = re.search(r'\((\d{4})\)$', query)
    year = year_match.group(1) if year_match else None
    if year:
        query = query.replace(f"({year})", "").strip()
    
    print(f"Query: {query}, Year: {year}")
    
    # Try API search first
    imdb_id = search_imdb_api(query, year)
    if imdb_id:
        print(f"Found via API: {imdb_id}")
        return imdb_id
    
    # Fall back to web scraping
    print("API search failed. Trying web scraping...")
    imdb_id = search_imdb_web(query, year)
    if imdb_id:
        print(f"Found via web scraping: {imdb_id}")
        return imdb_id
    
    print("No results found.")
    return None

class IMDbSearchGUI(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("IMDb Search")
        self.setGeometry(100, 100, 400, 300)

        central_widget = QWidget()
        self.setCentralWidget(central_widget)

        layout = QVBoxLayout()

        input_layout = QHBoxLayout()
        self.search_input = QLineEdit()
        self.search_input.returnPressed.connect(self.perform_search)  # Connect Enter key to search
        self.search_button = QPushButton("Search")
        self.search_button.clicked.connect(self.perform_search)
        input_layout.addWidget(self.search_input)
        input_layout.addWidget(self.search_button)

        self.result_display = QTextEdit()
        self.result_display.setReadOnly(True)

        layout.addLayout(input_layout)
        layout.addWidget(self.result_display)

        central_widget.setLayout(layout)

    def perform_search(self):
        query = self.search_input.text()
        self.result_display.clear()
        self.result_display.append(f"Searching for: {query}")
        
        imdb_id = search_imdb(query)
        
        if imdb_id:
            self.result_display.append(f"Final result: {imdb_id}")
            # Copy the movie ID to clipboard
            clipboard = QApplication.clipboard()
            clipboard.setText(imdb_id)
            self.result_display.append("Movie ID copied to clipboard!")
        else:
            self.result_display.append("No results found or an error occurred.")

if __name__ == "__main__":
    if "--gui" in sys.argv:
        app = QApplication(sys.argv)
        window = IMDbSearchGUI()
        window.show()
        sys.exit(app.exec())
    elif len(sys.argv) < 2:
        print("Usage: python imdb_search_gui.py <movie/show name> [(year)]")
        print("       python imdb_search_gui.py --gui")
        print("Example: python imdb_search_gui.py 'Gen V (2023)'")
        sys.exit(1)
    else:
        query = ' '.join(sys.argv[1:])
        imdb_id = search_imdb(query)
        
        if imdb_id:
            print(f"Final result: {imdb_id}")
            # Copy the movie ID to clipboard in CLI mode as well
            if sys.platform == "win32":
                import subprocess
                subprocess.run(["clip"], input=imdb_id.encode("utf-16"), check=True)
                print("Movie ID copied to clipboard!")
            elif sys.platform == "darwin":
                import subprocess
                subprocess.run("pbcopy", input=imdb_id.encode("utf-8"), check=True)
                print("Movie ID copied to clipboard!")
            elif sys.platform == "linux":
                import subprocess
                subprocess.run(["xclip", "-selection", "clipboard"], input=imdb_id.encode("utf-8"), check=True)
                print("Movie ID copied to clipboard!")
            else:
                print("Clipboard functionality not supported on this platform.")
        else:
            print("No results found or an error occurred.")

File: imdb_search.cpp
#include <iostream>
#include <string>
#include <regex>
#include <windows.h>
#include <wininet.h>
#include <sstream>
#include <algorithm>
#include "imdb_search_gui.h"

#pragma comment(lib, "wininet.lib")

std::string url_encode(const std::string &value) {
    std::string result;
    for (char c : value) {
        if (isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
            result += c;
        } else {
            char buf[4];
            sprintf_s(buf, "%%%02X", (unsigned char)c);
            result += buf;
        }
    }
    return result;
}

std::string http_get(const std::string& url) {
    std::string response;
    std::cout << "Sending request to: " << url << std::endl;
    HINTERNET hInternet = InternetOpenA("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36", INTERNET_OPEN_TYPE_DIRECT, NULL, NULL, 0);
    if (hInternet) {
        HINTERNET hConnect = InternetOpenUrlA(hInternet, url.c_str(), NULL, 0, INTERNET_FLAG_RELOAD, 0);
        if (hConnect) {
            char buffer[1024];
            DWORD bytesRead;
            while (InternetReadFile(hConnect, buffer, sizeof(buffer), &bytesRead) && bytesRead > 0) {
                response.append(buffer, bytesRead);
            }
            InternetCloseHandle(hConnect);
        } else {
            std::cout << "Failed to connect to URL. Error code: " << GetLastError() << std::endl;
        }
        InternetCloseHandle(hInternet);
    } else {
        std::cout << "Failed to initialize WinINet. Error code: " << GetLastError() << std::endl;
    }
    std::cout << "Response length: " << response.length() << std::endl;
    return response;
}

std::string search_imdb_api(const std::string& query, const std::string& year = "") {
    std::string encoded_query = url_encode(query);
    std::string search_url = "https://v3.sg.media-imdb.com/suggestion/x/" + encoded_query + ".json";
    
    std::string response = http_get(search_url);
    
    std::regex json_regex("\\{.*\\}");
    std::smatch json_match;
    if (std::regex_search(response, json_match, json_regex)) {
        std::string json_response = json_match.str();
        
        std::regex id_regex("\"id\":\"(tt\\d+)\"");
        std::regex title_regex("\"l\":\"([^\"]+)\"");
        std::regex year_regex("\"y\":(\\d{4})");
        
        std::sregex_iterator it(json_response.begin(), json_response.end(), id_regex);
        std::sregex_iterator end;
        
        while (it != end) {
            std::string id = (*it)[1];
            ++it;
            
            std::smatch title_match;
            if (std::regex_search(json_response, title_match, title_regex)) {
                std::string title = title_match[1];
                std::transform(title.begin(), title.end(), title.begin(), ::tolower);
                
                std::string query_lower = query;
                std::transform(query_lower.begin(), query_lower.end(), query_lower.begin(), ::tolower);
                
                if (title == query_lower) {
                    std::smatch year_match;
                    if (std::regex_search(json_response, year_match, year_regex)) {
                        std::string result_year = year_match[1];
                        if (year.empty() || result_year == year) {
                            return id;
                        }
                    }
                }
            }
        }
    }
    
    return "";
}

std::string search_imdb(const std::string& query) {
    std::cout << "Searching for: " << query << std::endl;
    
    std::regex year_regex("\\((\\d{4})\\)$");
    std::smatch year_match;
    std::string year;
    std::string cleaned_query = query;
    
    if (std::regex_search(query, year_match, year_regex)) {
        year = year_match[1];
        cleaned_query = query.substr(0, year_match.position());
        cleaned_query = cleaned_query.substr(0, cleaned_query.find_last_not_of(" \t") + 1);
    }
    
    std::cout << "Query: " << cleaned_query << ", Year: " << (year.empty() ? "Not specified" : year) << std::endl;
    
    std::string imdb_id = search_imdb_api(cleaned_query, year);
    if (!imdb_id.empty()) {
        std::cout << "Found via API: " << imdb_id << std::endl;
        return imdb_id;
    }
    
    std::cout << "No results found." << std::endl;
    return "";
}

void copy_to_clipboard(const std::string& text) {
    if (OpenClipboard(nullptr)) {
        EmptyClipboard();
        HGLOBAL hg = GlobalAlloc(GMEM_MOVEABLE, text.size() + 1);
        if (hg) {
            memcpy(GlobalLock(hg), text.c_str(), text.size() + 1);
            GlobalUnlock(hg);
            SetClipboardData(CF_TEXT, hg);
            CloseClipboard();
            std::cout << "Movie ID copied to clipboard!" << std::endl;
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: imdb_search.exe <movie/show name> [(year)]" << std::endl;
        std::cout << "       imdb_search.exe --gui" << std::endl;
        std::cout << "Example: imdb_search.exe \"The Shawshank Redemption (1994)\"" << std::endl;
        return 1;
    }
    
    if (std::string(argv[1]) == "--gui") {
        IMDbSearchGUI gui;
        gui.Show();
        return 0;
    }
    
    std::string query;
    for (int i = 1; i < argc; ++i) {
        query += argv[i];
        if (i < argc - 1) query += " ";
    }
    
    std::string imdb_id = search_imdb(query);
    
    if (!imdb_id.empty()) {
        std::cout << "Final result: " << imdb_id << std::endl;
        copy_to_clipboard(imdb_id);
    } else {
        std::cout << "No results found or an error occurred." << std::endl;
    }
    
    return 0;
}